mlflow/utils/checkpoint_utils.py

```diff
@@ -0,0 +1,174 @@
+
+
+class _MlflowModelCheckpointCallbackBase(metaclass=ExceptionSafeAbstractClass):
+
+    def __init__(
+        self,
+        checkpoint_file_suffix,
+        monitor,
+        mode,
+        save_best_only,
+        save_weights_only,
+        save_freq,
+    ):
+        """
+        Args:
+            checkpoint_file_suffix: checkpoint file suffix.
+            monitor: In automatic model checkpointing, the metric name to monitor if
+                you set `model_checkpoint_save_best_only` to True.
+            save_best_only: If True, automatic model checkpointing only saves when
+                the model is considered the "best" model according to the quantity
+                monitored and previous checkpoint model is overwritten.
+            mode: one of {"min", "max"}. In automatic model checkpointing,
+                if save_best_only=True, the decision to overwrite the current save file is made
+                based on either the maximization or the minimization of the monitored quantity.
+            save_weights_only: In automatic model checkpointing, if True, then
+                only the model’s weights will be saved. Otherwise, the optimizer states,
+                lr-scheduler states, etc are added in the checkpoint too.
+            save_freq: `"epoch"` or integer. When using `"epoch"`, the callback
+                saves the model after each epoch. When using integer, the callback
+                saves the model at end of this many batches. Note that if the saving isn't
+                aligned to epochs, the monitored metric may potentially be less reliable (it
+                could reflect as little as 1 batch, since the metrics get reset
+                every epoch). Defaults to `"epoch"`.
+        """
+        self.checkpoint_file_suffix = checkpoint_file_suffix
+        self.monitor = monitor
+        self.mode = mode
+        self.save_best_only = save_best_only
+        self.save_weights_only = save_weights_only
+        self.save_freq = save_freq
+        self.last_monitor_value = None
+
+        if self.save_best_only:
+            if self.monitor is None:
+                raise MlflowException(
+                    "If checkpoint 'save_best_only' config is set to True, you need to set "
+                    "'monitor' config as well."
+                )
+            if self.mode not in ["min", "max"]:
+                raise MlflowException(
+                    "If checkpoint 'save_best_only' config is set to True, you need to set "
+                    "'mode' config and available modes includes 'min' and 'max', but you set "
+                    f"'mode' to '{self.mode}'."
+                )
+
+    def _is_new_checkpoint_better(self, new_monitor_value):
+        if self.last_monitor_value is None:
+            return True
+
+        if self.mode == "min":
+            return new_monitor_value <= self.last_monitor_value
+
+        return new_monitor_value >= self.last_monitor_value
```

Comment:
Do we need `=` here?

Reply:
Let's say you have a weight-less model. No chance for this model to improve. In the current logic, a new checkpoint is logged in each epoch, but that seems useless.